### ISSUE TITLE
Client – Figure out how to handle lossy state conversions between RGB/Hex to HSV

### ISSUE THREAD

**Issue by Parkreiner** [Parkreiner/etrian-character-customizer#8]
Already done; just adding as project task to document it.

**Comment by Parkreiner** [comment]
Adding some notes to document why state management was so screwy here:
1. One of the design goals I had for the app was that I wanted to have the majority of the app be based around hex color strings. This would have two main benefits:
   - The strings would be easier to send over from the server, as compared to a ton of objects
   - The strings would work better with React's reference-based model, making any memoization easier down the line, and making it a lot easier to prevent any `useEffect` dependency arrays from over-running.
2. Of course, working with hex strings is awkward, so I decided there would be a single `ColorPicker` component that would deal with taking the hex strings and converting them into RGB/HSV values. Only this one component would ever need to know that these other color formats even exist.
3. To make it clear what options are available at all times, I wanted to display both RGB and HSV values at the same time. The problem is, hex to RGB is perfectly fine, but RGB to HSV is a lossy conversion process, meaning that it doesn't line up with the general React approach of "derive values as much as possible".
   - I think there's a reason why a lot of React-based color editors (including Adobe Color) don't let you see both at the same time. They're not counting on you noticing that values can jump around a little erratically, just because of how the math formulas work, and general rounding issues.

So basically, the solution was to have the component treat the HSV value as state and keep track of the last hex value that the component dispatched. Whenever the component would re-render, it would compare the hex value received through props against the last known hex, and force a state sync if they didn't match.
